exercises/tree-building/examples/success-standard/src/TreeBuilding.hs

```diff
@@ -0,0 +1,77 @@
+module TreeBuilding (newTree, Record(..), Tree(..)) where
+
+import Data.List
+import Data.Maybe (isNothing)
+
+type Id = Int
+type Children = [Tree]
+type ParentGrouping = (Maybe Id, [Id])
+
+data Record = Record Id (Maybe Id) deriving (Eq, Show)
+
+data Tree = Leaf Id | Branch Id Children deriving (Eq, Show)
+
+newTree :: [Record] -> Maybe Tree
+newTree records
+  | cycles records      = Nothing
+  | succIdCheck records = build . groupByParent $ records
+  | otherwise           = Nothing
+
+-- Checks for cycling in newTree
+cycles :: [Record] -> Bool
+cycles = any checkRecord
+  where checkRecord r =
+          case r of
+            (Record i Nothing)  -> if' (i /= 0) True False
+            (Record i (Just p)) -> if' (p >= i) True False
+
+build :: [ParentGrouping] -> Maybe Tree
+build [] = Nothing
+build (x:xs)
+  | not . validRoot $ x = Nothing
+  | null xs             = Just (Leaf 0)
+  | otherwise           = Just (Branch 0 (build' xs))
+
+--  Internal build
+build' :: [ParentGrouping] -> [Tree]
+build' xs
+  | length xs == 1 = map Leaf (snd (head xs))
+  | otherwise      = buildWithChildren xs
+  where
+    buildWithChildren (y:ys) =
+      map (\yid ->
+             let children = filter ((== Just yid) . fst) ys
+             in if not (null children)
+                then Branch yid (build' children)
+                else Leaf yid
+          ) (snd y)
+
+-- Validate the Root node
+validRoot :: ParentGrouping -> Bool
+validRoot (p, ids) = isNothing p && length ids == 1 && sum ids == 0
+
+groupByParent :: [Record] -> [ParentGrouping]
+groupByParent = sortOn fst
+              . map (\xs -> (recordParent (head xs), map recordId xs))
```

Comment:
```haskell
import Control.Arrow ((&&&))

(\x -> (f x, g x)) = f &&& g
```
```suggestion
              . map (recordParent . head &&& map recordId))
```

Reply:
You may like to know that `first` and `second` actually live in two places nowadays: There is a better version in [`Data.Bifunctor`](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Bifunctor.html#v:first). A bifunctor is considerably less scary than an arrow, I think.